odxtools/diagcodedtype.py

```diff
@@ -89,46 +94,45 @@ def _extract_internal(
         cursor_position = byte_position + byte_length
         extracted_bytes = coded_message[byte_position:cursor_position]
 
-        # Apply byteorder
-        if not is_highlow_byte_order and base_data_type not in [
-                DataType.A_UNICODE2STRING,
-                DataType.A_BYTEFIELD,
-                DataType.A_ASCIISTRING,
-                DataType.A_UTF8STRING,
+        # Apply byteorder for numerical objects. Note that doing this
+        # here might lead to garbage data being included in the result
+        # if the data to be extracted is not byte aligned and crosses
+        # byte boundaries, but it is what the specification says.
+        if not is_highlow_byte_order and base_data_type in [
+                DataType.A_INT32,
+                DataType.A_UINT32,
+                DataType.A_FLOAT32,
+                DataType.A_FLOAT64,
         ]:
             extracted_bytes = extracted_bytes[::-1]
 
-        format_letter = ODX_TYPE_TO_FORMAT_LETTER[base_data_type]
-        padding = 8 * byte_length - (bit_length + bit_position)
-        text_encoding = 'utf-8'
-        text_errors = 'strict'
-        if not exceptions.strict_mode:
-            text_errors = 'replace'
+        padding = (8 - (bit_length + bit_position) % 8) % 8
+        internal_value, = bitstruct.unpack_from(
```

Comment:
Consider compiling the bitstruct format using `bitstruct.compile` and cache the result in the codedtype itself, for a significant performance boost

Reply:
I played with this thought, but concluded that it is not worth the effort: it would imply that the arguments of the `_extract_internal_value()` function must be baked into the `CompiledFormat` object (i.e., a major redesign of code that uses this function would be required), and second, for `bitstruct.c` the unpack operation with these simple formats should be about the same speed using compiled objects as for the uncompiled ones because I suspect that performance is dominated by the overhead of the python interpreter not by the execution of the function itself (I haven't benchmarked it, though...)

Reply:
okay, I benchmarked it a bit: using `bitstruct.c` instead of `bitstruct` to unpack the format `"p4u8"` comes with a speedup of about factor 30; using a precompiled format object with `bitstruct.c` yields another speedup of roughly factor `1.5`. That said, I am pretty certain that the coding functions are not dominated by the packing and unpacking routines when using `bitstruct.c`...